### PR TITLE
dts: bindings: add XIP node for nordic devices

### PR DESCRIPTION
**Pull request by hakonfam**
There is currently no information available about the XIP
addresses for the nRF52840 and nRF5340.

Add a new node type and add the nodes to the qspi bus.

Signed-off-by: Håkon Øye Amundsen <haakon.amundsen@nordicsemi.no>

### CODE REVIEW

**Comment by mbolivar-nordic** [comment]
@galak I am curious how you think this sort of thing should be handled as I expect other SoCs will need similar or may already have similar nodes.

We are trying to model this XIP area of a memory map:

Screenshot from 2021-08-31 13-24-21

This is memory mapped external flash memory and the flash controller in this case is the nordic,nrf-qspi node.

**Comment by galak** [comment]
@mbolivar-nordic at a quick glance this seems similar to some other QSPI devices that expose a memory map XIP region.  We use the reg property of the qspi and have one of the reg's map to the XIP region.

Here's an example:
https://git.kernel.org/pub/scm/linux/kernel/git/devicetree/devicetree-rebasing.git/tree/Bindings/spi/ti_qspi.txt

File: dts/bindings/mtd/nordic,nrf-qspi-xip.yaml [hunk 1]
```diff
@@ -0,0 +1,13 @@
+# Copyright (c) 2021 Nordic Semiconductor ASA
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
+  FOR INTERNAL USE ONLY
```
Comment (galak) [inline dts/bindings/mtd/nordic,nrf-qspi-xip.yaml:5]: what is this meant to me?

**Review by galak — CHANGES_REQUESTED** [review]
Need to fix description in binding.

**Comment by galak** [comment]
Is it really the case that there is 1 byte that isn't part of the XIP region?  That seems very weird.

**Comment by mbolivar-nordic** [comment]
> Need to fix description in binding.

Thanks, yeah. We probably should have tagged this RFC for now; it was more a question about how to represent this than about the details in the description.

**Comment by mbolivar-nordic** [comment]
> the reg property of the qspi and have one of the reg's map to the XIP region.
> 
> Here's an example:
> https://git.kernel.org/pub/scm/linux/kernel/git/devicetree/devicetree-rebasing.git/tree/Bindings/spi/ti_qspi.txt

That looks clean, especially with the reg names.

**Comment by mbolivar-nordic** [comment]
@galak I think this is what you meant; right? I had to extend the cmake helpers to make this useful.

**Comment by mbolivar-nordic** [comment]
@tejlmand @hakonfam example:

```cmake
dt_nodelabel(qspi_path NODELABEL "qspi")

dt_reg_addr(qspi_addr PATH ${qspi_path} NAME qspi)
dt_reg_size(qspi_size PATH ${qspi_path} NAME qspi)
dt_reg_addr(qspi_mm_addr PATH ${qspi_path} NAME qspi_mm)
dt_reg_size(qspi_mm_size PATH ${qspi_path} NAME qspi_mm)

message(STATUS "qspi_addr=${qspi_addr}")
message(STATUS "qspi_mm_addr=${qspi_mm_addr}")
message(STATUS "qspi_size=${qspi_size}")
message(STATUS "qspi_mm_size=${qspi_mm_size}")
```

output on nRF52840:

```
-- qspi_addr=0x40029000
-- qspi_mm_addr=0x12000000
-- qspi_size=0x1000
-- qspi_mm_size=0xdffffff
```

> Is it really the case that there is 1 byte that isn't part of the XIP region? That seems very weird.

@hakonfam could you check on this?

**Comment by mbolivar-nordic** [comment]
> cmake: extensions: fix copy/paste issues in DT error messages

I split this part into a separate PR for LTS: https://github.com/zephyrproject-rtos/zephyr/pull/38185

**Comment by hakonfam** [comment]
> Is it really the case that there is 1 byte that isn't part of the XIP region? That seems very weird.

I fixed that.

File: cmake/extensions.cmake [hunk 2]
```diff
@@ -2760,6 +2780,17 @@ function(dt_reg_size var)
   set(${var} ${${var}} PARENT_SCOPE)
 endfunction()
 
+# Internal helper for dt_reg_addr/dt_reg_size; not meant to be used directly
+function(_dt_reg_get_index var path name)
```
Comment (tejlmand) [inline cmake/extensions.cmake:2784]: Don't use `_` as first character of a function.
The `_` is used by CMake to prepend existing functions when they are overloaded, as can be seen by this example:
```
# test.cmake
function(A)
  message("I'm in A")
endfunction()

function(A)
  message("I'm in B")
endfunction()

A()
_A()
```
and running this:
```
$ cmake -P test.cmake
I'm in B
I'm in A
```

**Review by tejlmand — CHANGES_REQUESTED** [review]
Please, don't use `_` as first character in a function name, that's bad habit in CMake.

File: cmake/extensions.cmake (hunk 2 above)
Comment (mbolivar-nordic) [reply cmake/extensions.cmake]: > `_` is used by CMake to prepend existing functions when they are overloaded

Cool. I didn't know that. Fixed now.

File: cmake/extensions.cmake [hunk 3]
```diff
@@ -2760,6 +2780,17 @@ function(dt_reg_size var)
...
 
+# Internal helper for dt_reg_addr/dt_reg_size; not meant to be used directly
+function(dt_reg_index_private var path name)
+  dt_prop(reg_names PATH "${path}" PROPERTY "reg-names")
+  if(NOT DEFINED reg_names)
+    set(index "-1")
+  else()
+    list(FIND reg_names "${name}" index)
+  endif()
+  set("${var}" "${index}" PARENT_SCOPE)
```
Comment (tejlmand) [inline cmake/extensions.cmake:2791]: Why the quotes on the return var ?

If there is any risk that `${var}` can be undefined or empty string, then we should not try to do a `set()` on it, but based on how this is used, that is not really a risk.
```suggestion
  set(${var} "${index}" PARENT_SCOPE)
```

**Review by tejlmand — CHANGES_REQUESTED** [review]

**Comment by mbolivar-nordic** [comment]
Done

File: cmake/extensions.cmake (hunk 3 above)
Comment (tejlmand) [reply cmake/extensions.cmake]: @mbolivar-nordic why was this marked resolved without any comments or fixes ?

File: cmake/extensions.cmake [hunk 4]
```diff
@@ -2788,7 +2788,7 @@ function(dt_reg_index_private var path name)
   else()
     list(FIND reg_names "${name}" index)
   endif()
-  set("${var}" "${index}" PARENT_SCOPE)
+  set(${var} "${index}" PARENT_SCOPE)
```
Comment (tejlmand) [inline cmake/extensions.cmake:2791]: ahh, here's the reason.
I believe this should be in previous commit.

Comment (mbolivar-nordic) [reply cmake/extensions.cmake:2791]: > I believe this should be in previous commit.

@tejlmand You're right, of course. Fixed.

**Review by tejlmand — APPROVED** [review]
lgtm.

**Comment by mbolivar-nordic** [comment]
Rebased

**Comment by cfriedt** [comment]
@galak - please go ahead and merge if the requested changes have been made (quite likely, I think).

**Review by galak — APPROVED** [review]

**Pull request merged**